File: results/2026-05-06_173435/tasks/15-test-results-aggregator/typescript-bun-opus47-1m-high/generated-code/scripts/run-act-tests.ts
// End-to-end harness that drives the workflow through nektos/act.
//
// Each test case:
//   1. Materialises a temp git repo with the project files.
//   2. Replaces the `fixtures/` tree with case-specific data.
//   3. Runs `act push --rm` and captures the output.
//   4. Appends the output to `act-result.txt` with clear delimiters.
//   5. Asserts the EXACT expected RESULTS line and that every job succeeded.
//
// The full unit-test suite still runs inside each case, since the workflow
// invokes `bun test` before aggregating. That satisfies the "every test case
// runs through act" rule without spending an act run per unit test.
//
// Run: `bun run scripts/run-act-tests.ts` (≈30–90s per case).

import {
  cpSync,
  mkdtempSync,
  rmSync,
  writeFileSync,
  appendFileSync,
  existsSync,
  readdirSync,
  statSync,
  mkdirSync,
} from "node:fs";
import { tmpdir } from "node:os";
import { join, resolve } from "node:path";
import { spawnSync, type SpawnSyncReturns } from "node:child_process";

interface FixtureFile {
  path: string; // relative to repo root, e.g. "fixtures/junit/x.xml"
  content: string;
}

interface ActCase {
  name: string;
  fixtures: FixtureFile[];
  // Substring assertions on the captured output. Each must appear at least once.
  expectContains: string[];
  // Substring exclusions — must NOT appear. Used to lock in negative cases.
  expectMissing?: string[];
}

const REPO_ROOT = resolve(import.meta.dir, "..");
const ACT_RESULT_PATH = join(REPO_ROOT, "act-result.txt");

// Case 1: the bundled fixture set — 4 files, 16 tests, 2 failures, 2 flaky.
const CASE_BUNDLED: ActCase = {
  name: "bundled-fixtures",
  fixtures: [], // empty → use the repo's own fixtures/ unchanged
  expectContains: [
    "RESULTS files=4 total=16 passed=13 failed=2 skipped=1 flaky=2 duration=4.60s",
    "FLAKY payments.CheckoutTests::handles_3ds_redirect,search.IndexTests::ranks_by_relevance",
    "Job succeeded",
  ],
};

// Case 2: an all-green run with a single JSON file. Asserts on different exact
// numbers so we know the harness isn't just matching boilerplate.
const CASE_ALL_GREEN: ActCase = {
  name: "all-green-single-file",
  fixtures: [
    {
      path: "fixtures/json/all-green.json",
      content: JSON.stringify({
        suite: "smoke.Suite",
        tests: [
          { name: "boots", classname: "smoke.Suite", status: "passed", duration: 0.5 },
          { name: "responds_to_ping", classname: "smoke.Suite", status: "passed", duration: 0.25 },
          { name: "writes_logs", classname: "smoke.Suite", status: "passed", duration: 0.15 },
        ],
      }),
    },
  ],
  expectContains: [
    "RESULTS files=1 total=3 passed=3 failed=0 skipped=0 flaky=0 duration=0.90s",
    "Status: all tests passed",
    "Job succeeded",
  ],
  expectMissing: ["FLAKY ", "## Failures"],
};

const CASES: ActCase[] = [CASE_BUNDLED, CASE_ALL_GREEN];

function copyProjectInto(dest: string): void {
  // Copy everything except .git, node_modules, and build artifacts.
  const skip = new Set([".git", "node_modules", "act-result.txt"]);
  for (const entry of readdirSync(REPO_ROOT)) {
    if (skip.has(entry)) continue;
    const src = join(REPO_ROOT, entry);
    const dst = join(dest, entry);
    cpSync(src, dst, { recursive: true });
  }
}

function rmDir(dir: string): void {
  if (existsSync(dir)) rmSync(dir, { recursive: true, force: true });
}

function applyFixtures(repoDir: string, fixtures: FixtureFile[]): void {
  if (fixtures.length === 0) return; // keep bundled fixtures
  // Wipe and rewrite the fixtures dir so cases see only their own data.
  rmDir(join(repoDir, "fixtures"));
  for (const f of fixtures) {
    const full = join(repoDir, f.path);
    mkdirSync(full.substring(0, full.lastIndexOf("/")), { recursive: true });
    writeFileSync(full, f.content);
  }
}

function gitInit(repoDir: string): void {
  // act needs a git repo to detect the workflow event context.
  const run = (args: string[]) => {
    const r = spawnSync("git", args, { cwd: repoDir, encoding: "utf8" });
    if (r.status !== 0) {
      throw new Error(
        `git ${args.join(" ")} failed: ${r.stderr || r.stdout}`,
      );
    }
  };
  run(["init", "-q", "-b", "main"]);
  run(["config", "user.email", "harness@example.com"]);
  run(["config", "user.name", "harness"]);
  run(["add", "-A"]);
  run(["commit", "-q", "-m", "harness fixture"]);
}

function runAct(repoDir: string): SpawnSyncReturns<string> {
  // --rm: remove containers after run. -P: image override taken from .actrc.
  return spawnSync("act", ["push", "--rm"], {
    cwd: repoDir,
    encoding: "utf8",
    maxBuffer: 50 * 1024 * 1024,
  });
}

function appendCaseOutput(name: string, output: string, exitCode: number | null): void {
  const banner =
    `\n${"=".repeat(80)}\n` +
    `CASE: ${name}\n` +
    `EXIT: ${exitCode}\n` +
    `${"=".repeat(80)}\n`;
  appendFileSync(ACT_RESULT_PATH, banner + output + "\n");
}

function assertContains(output: string, needles: string[], caseName: string): void {
  for (const n of needles) {
    if (!output.includes(n)) {
      throw new Error(`[${caseName}] expected output to contain: ${JSON.stringify(n)}`);
    }
  }
}

function assertMissing(output: string, needles: string[], caseName: string): void {
  for (const n of needles) {
    if (output.includes(n)) {
      throw new Error(`[${caseName}] expected output to NOT contain: ${JSON.stringify(n)}`);
    }
  }
}

function countOccurrences(haystack: string, needle: string): number {
  let count = 0;
  let i = 0;
  while ((i = haystack.indexOf(needle, i)) !== -1) {
    count++;
    i += needle.length;
  }
  return count;
}

async function main(): Promise<void> {
  // Truncate the result file so re-runs don't accumulate stale output.
  writeFileSync(ACT_RESULT_PATH, `act test harness — ${new Date().toISOString()}\n`);

  let failures = 0;
  for (const c of CASES) {
    const dir = mkdtempSync(join(tmpdir(), `act-${c.name}-`));
    try {
      copyProjectInto(dir);
      applyFixtures(dir, c.fixtures);
      gitInit(dir);

      console.log(`[${c.name}] running act in ${dir} ...`);
      const r = runAct(dir);
      const output = (r.stdout ?? "") + (r.stderr ?? "");
      appendCaseOutput(c.name, output, r.status);

      if (r.status !== 0) {
        console.error(`[${c.name}] act exited with code ${r.status}`);
        failures++;
        continue;
      }
      try {
        assertContains(output, c.expectContains, c.name);
        if (c.expectMissing) assertMissing(output, c.expectMissing, c.name);
        // act prints "Job succeeded" once per job; we only have one job, so
        // exactly one occurrence is expected.
        const succeeded = countOccurrences(output, "Job succeeded");
        if (succeeded < 1) {
          throw new Error(`[${c.name}] no "Job succeeded" line found`);
        }
        console.log(`[${c.name}] PASS`);
      } catch (err) {
        console.error((err as Error).message);
        failures++;
      }
    } finally {
      rmDir(dir);
    }
  }

  if (failures > 0) {
    console.error(`\n${failures} of ${CASES.length} act case(s) FAILED`);
    process.exit(1);
  }
  console.log(`\nAll ${CASES.length} act case(s) PASSED — see act-result.txt`);
}

await main();

File: results/2026-05-06_173435/tasks/15-test-results-aggregator/default-opus47-1m-medium/generated-code/run_act_tests.py
#!/usr/bin/env python3
"""Test harness: run the workflow under `act` for each fixture-set test case.

Every assertion exercises the aggregator *through the GitHub Actions workflow*
(per task requirements) — we never call aggregator.py directly here.

For each case, we:
  1. Build a temp git repo containing the project + only that case's fixture files
  2. Run `act push --rm` once (capped at 3 total runs to stay under the budget)
  3. Append all output to act-result.txt
  4. Assert act exited 0, every job says "Job succeeded", and the STATS line
     printed by aggregator.py contains the exact expected counts for this case.
"""
from __future__ import annotations

import re
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

PROJECT = Path(__file__).resolve().parent
ACT_RESULT = PROJECT / "act-result.txt"

PROJECT_FILES = [
    "aggregator.py",
    "test_aggregator.py",
    ".actrc",
    ".github/workflows/test-results-aggregator.yml",
]

# Each case: name -> {fixtures (list of source filenames), expected stats}.
# Expected values were hand-computed from the fixture contents — see fixtures/.
CASES = [
    {
        "name": "all_three_runs",
        "fixtures": ["run1.xml", "run2.json", "run3.xml"],
        # run1: 3p/1f/1s, run2: 4p/0f/1s, run3: 1p/1f/0s
        # token_refresh failed in run1, passed in run2 → flaky
        # charge_card failed only (consistently) → real failure
        "expected": {"passed": 8, "failed": 2, "skipped": 2, "flaky": 1,
                     "duration": 6.43},
        "expect_md": ["payments.ChargeTests.test_charge_card",
                      "auth.LoginTests.test_token_refresh"],
    },
    {
        "name": "json_only_clean",
        "fixtures": ["run2.json"],
        "expected": {"passed": 4, "failed": 0, "skipped": 1, "flaky": 0,
                     "duration": 2.83},
        "expect_md": ["All tests passed cleanly"],
    },
    {
        "name": "single_failing_xml",
        "fixtures": ["run3.xml"],
        "expected": {"passed": 1, "failed": 1, "skipped": 0, "flaky": 0,
                     "duration": 0.60},
        "expect_md": ["payments.ChargeTests.test_charge_card"],
    },
]


def stage_repo(case: dict) -> Path:
    """Create a temp git repo containing the project + only this case's fixtures."""
    workdir = Path(tempfile.mkdtemp(prefix=f"act-{case['name']}-"))
    for rel in PROJECT_FILES:
        src = PROJECT / rel
        dst = workdir / rel
        dst.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(src, dst)
    fixtures_dir = workdir / "fixtures"
    fixtures_dir.mkdir()
    for f in case["fixtures"]:
        shutil.copy2(PROJECT / "fixtures" / f, fixtures_dir / f)
    # Initialize git so act has a sha to operate on.
    subprocess.run(["git", "init", "-q", "-b", "main"], cwd=workdir, check=True)
    subprocess.run(["git", "config", "user.email", "t@t"], cwd=workdir, check=True)
    subprocess.run(["git", "config", "user.name", "t"], cwd=workdir, check=True)
    subprocess.run(["git", "add", "."], cwd=workdir, check=True)
    subprocess.run(["git", "commit", "-q", "-m", "init"], cwd=workdir, check=True)
    return workdir


def run_act(workdir: Path) -> tuple[int, str]:
    proc = subprocess.run(
        ["act", "push", "--rm", "--pull=false"],
        cwd=workdir,
        capture_output=True,
        text=True,
        timeout=600,
    )
    return proc.returncode, proc.stdout + proc.stderr


def parse_stats(output: str) -> dict | None:
    """Find the `STATS passed=X failed=Y ...` line emitted by aggregator.py."""
    m = re.search(
        r"STATS passed=(\d+) failed=(\d+) skipped=(\d+) flaky=(\d+) duration=([\d.]+)",
        output,
    )
    if not m:
        return None
    return {
        "passed": int(m.group(1)),
        "failed": int(m.group(2)),
        "skipped": int(m.group(3)),
        "flaky": int(m.group(4)),
        "duration": float(m.group(5)),
    }


def main() -> int:
    ACT_RESULT.write_text("")  # truncate at start
    failures: list[str] = []

    for case in CASES:
        print(f"\n=== CASE: {case['name']} ===", flush=True)
        workdir = stage_repo(case)
        try:
            rc, output = run_act(workdir)
        finally:
            # Always preserve output even on exceptions
            with ACT_RESULT.open("a") as fh:
                fh.write(f"\n\n========== CASE: {case['name']} ==========\n")
                fh.write(f"workdir: {workdir}\n")
                fh.write(output if 'output' in locals() else "<no output>\n")
                fh.write(f"\n--- exit code: {rc if 'rc' in locals() else '?'} ---\n")

        # --- Assertions ---
        if rc != 0:
            failures.append(f"{case['name']}: act exit code {rc} (expected 0)")
            continue

        # Every job in act prints either "Job succeeded" or "Job failed". Require all succeeded.
        if "Job failed" in output:
            failures.append(f"{case['name']}: at least one job failed")
        if "Job succeeded" not in output:
            failures.append(f"{case['name']}: no 'Job succeeded' line found")

        stats = parse_stats(output)
        if stats is None:
            failures.append(f"{case['name']}: STATS line not found in output")
            continue

        exp = case["expected"]
        for key in ("passed", "failed", "skipped", "flaky"):
            if stats[key] != exp[key]:
                failures.append(
                    f"{case['name']}: {key} = {stats[key]}, expected {exp[key]}"
                )
        # Allow tiny float slop on duration
        if abs(stats["duration"] - exp["duration"]) > 0.01:
            failures.append(
                f"{case['name']}: duration = {stats['duration']}, "
                f"expected ≈ {exp['duration']}"
            )

        for needle in case["expect_md"]:
            if needle not in output:
                failures.append(
                    f"{case['name']}: expected substring not in output: {needle!r}"
                )

        # Tidy up the temp dir on success
        shutil.rmtree(workdir, ignore_errors=True)

    print("\n========== HARNESS RESULT ==========")
    if failures:
        for f in failures:
            print(f"FAIL: {f}")
        print(f"\n{len(failures)} assertion(s) failed.")
        return 1
    print(f"All {len(CASES)} cases passed.")
    print(f"Full act output captured in: {ACT_RESULT}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
